src/v/cloud_roles/refresh_credentials.cc

```diff
@@ -0,0 +1,247 @@
+              _api_host,
+              *host_override);
+            _api_host = {*host_override};
+        }
+        if (auto port_override = load_and_validate_env_var(
+              override_api_endpoint_env_vars::port);
+            port_override) {
+            try {
+                auto override_port = std::stoi(*port_override);
+                vlog(
+                  clrl_log.debug,
+                  "api_port overridden from {} to {}",
+                  _api_port,
+                  override_port);
+                _api_port = override_port;
+            } catch (...) {
+                vlog(
+                  clrl_log.error,
+                  "failed to convert port value {} from string to uint16_t",
+                  *port_override);
+            }
+        }
+    }
+}
+
+ss::future<> cloud_roles::refresh_credentials::fetch_and_update_credentials() {
+    // Before fetching new credentials, either:
+    // 1. do not sleep - this is an initial call to API
+    // 2. sleep until we are close to expiry of credentials
+    // 3. sleep in case of retryable failure for a short duration
+
+    co_await sleep_until_expiry();
+
+    vlog(clrl_log.debug, "fetching credentials");
+    auto fetch_response = co_await fetch_credentials();
+    auto handle_result = handle_response(std::move(fetch_response));
+
+    co_return co_await ss::visit(
+      std::move(handle_result),
+      [this](malformed_api_response_error err) {
+          _probe.fetch_failed();
+          vassert(
+            err.missing_fields.empty(),
+            "bad api response, missing fields: {}",
+            err.missing_fields);
+          return ss::now();
+      },
+      [this](api_response_parse_error err) {
+          _probe.fetch_failed();
+          vlog(clrl_log.error, "failed to parse api response: {}", err.reason);
+          return ss::now();
+      },
+      [this](api_request_error err) {
+          _probe.fetch_failed();
+          switch (err.error_kind) {
+          case api_request_error_kind::failed_abort:
+              vassert(false, "api request failed: {}", err.reason);
+          case api_request_error_kind::failed_retryable:
+              // handle_response has set the sleep duration, we will retry after
+              // a cool-off period
+              vlog(
+                clrl_log.error,
+                "api request failed (retrying after cool-off period): {}",
+                err.reason);
+          }
+          return ss::now();
+      },
+      [this](credentials creds) {
+          _probe.fetch_success();
+          vlog(clrl_log.info, "fetched credentials {}", creds);
+          return _credentials_update(std::move(creds));
+      });
+}
+
+ss::lowres_clock::duration
+cloud_roles::refresh_credentials::impl::calculate_sleep_duration(
+  uint32_t expiry_sec) const {
+    int sleep = std::floor(expiry_sec * sleep_from_expiry_multiplier);
+    return std::chrono::seconds{sleep};
+}
+
+ss::lowres_clock::duration
```

Comment:
nit: it ok to use `ss::lowres_clock::duration` but it is a little more idiomatic to use standard types like std::chrono::milliseconds and then convert to other duration types when needed (for example when settings a timer based on a lowres_clock).